pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConnectionPool.java

```diff
@@ -140,12 +141,23 @@ public void initChannel(SocketChannel ch) throws Exception {
 
             future.channel().closeFuture().addListener(v -> {
                 // Remove connection from pool when it gets closed
+                if (log.isDebugEnabled()) {
+                    log.debug("Removing closed connection from pool: {}", v);
+                }
                 cleanupConnection(address, connectionKey, cnxFuture);
             });
 
             // We are connected to broker, but need to wait until the connect/connected handshake is
             // complete
             final ClientCnx cnx = (ClientCnx) future.channel().pipeline().get("handler");
+            if (!future.channel().isActive() || cnx == null) {
```

Comment:
future.isSuccess() = true implies that future.channel() is active. That's guaranteed right? It should be but just wanted to confirm.